### PR TITLE
Add FieldArrayWithDefaults and WaveformArray to record.py

### PR DESCRIPTION
**Pull request by cdcapano**
This adds a base class _FieldArrayWithDefaults and WaveformArray to record.py. WaveformArray is an instance of FieldArray with some common default fields and methods defined. A subset of these fields can be set up on initialization using the names keyword argument. The names can include default fields, virtual fields, methods, and/or functions of these. If a virtual field or method is included in the names, the underlying source code is analyzed so that all the necessary fields are included for the virtual field or method to work. See the doc string for WaveformArray for examples.

I intend to use this in the inference plotting codes, to make plotting different parameters easier. (Pull request to follow this one.)

### CODE REVIEW

File: pycbc/io/record.py [hunk 1]
```diff
@@ -967,4 +1041,353 @@ def add_fields(self, arrays, names=None, assubarray=False):
...
+#
+# =============================================================================
+#
+#                           WaveformArray
+#
+# =============================================================================
+#
+
+# we'll vectorize TimeDelayFromEarthCenter for faster processing of end times
+def _time_delay_from_center(tc, detector, ra, dec):
```
Comment (ahnitz) [inline pycbc/io/record.py]: Please use the pycbc.detector.Detector class instead of this function.

File: pycbc/io/record.py [hunk 2]
```diff
@@ -967,4 +1041,353 @@ def add_fields(self, arrays, names=None, assubarray=False):
...
+    ('foo',)    
+
+    Additional fields can also be specified using the additional_fields
+    keyword argument. Example:
+    >>> warr = WaveformArray(10, names=['mass1', 'mass2'], additional_fields=[('lambda1', float)])
+    >>> warr.fieldnames
+    ('mass1', 'mass2', 'lambda1')
+    """
+
+    _staticfields = {
```
Comment (ahnitz) [inline pycbc/io/record.py]: These fields should not be built in here. Can you please read them from the pycbc.wavefom.waveform module.

File: pycbc/io/record.py [hunk 3]
```diff
@@ -967,4 +1041,353 @@ def add_fields(self, arrays, names=None, assubarray=False):
...
+            'ra': float,
+            'dec': float,
+            'polarization': float,
+            'inclination': float,
+    }
+
+    _virtualfields = ['mP', 'mS', 'mtotal', 'q', 'eta', 'mchirp']
+
+    @property
+    def mP(self):
```
Comment (ahnitz) [inline pycbc/io/record.py]: the names of these methods violate pep8 , please make sure they don't use capitals. https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables

**Comment by cdcapano** [comment]
Ok, I've added a new module to waveform called parameters. This has classes Parameter and ParameterList, which allows you to define parameters with extra meta data, such as dtype, description, etc. I then define a bunch of parameters that we typically use, and construct lists of these parameters for use by other codes. In waveform, `default_args` now pulls from these parameter lists, and the doc strings for the various `get_*_waveform*` functions use the parameter lists. The list of parameters is now used for the default fields in WaveformArray.

I checked that the `default_args` in waveform matches what it was before. I also checked that the doc strings for the ``get_*_waveform*`` functions make sense. Here, e.g., is what I get for ``get_fd_waveform``:
```
Return a frequency domain gravitational waveform.`

Parameters
----------
template: object
    An object that has attached properties. This can be used to substitute
    for keyword arguments. A common example would be a row in an xml table.
mass1 : {None, float}
    The mass of the first component object in the binary (in solar masses).
mass2 : {None, float}
    The mass of the second component object in the binary (in solar masses).
spin1x : {0.0, float}
    The x component of the first binary component's dimensionless spin.
spin1y : {0.0, float}
    The y component of the first binary component's dimensionless spin.
spin1z : {0.0, float}
    The z component of the first binary component's dimensionless spin.
spin2x : {0.0, float}
    The x component of the second binary component's dimensionless spin.
spin2y : {0.0, float}
    The y component of the second binary component's dimensionless spin.
spin2z : {0.0, float}
    The z component of the second binary component's dimensionless spin.
lambda1 : {0.0, float}
    The tidal deformability parameter of object 1.
lambda2 : {0.0, float}
    The tidal deformability parameter of object 2.
distance : {1.0, float}
    Luminosity distance to the binary (in Mpc).
coa_phase : {0.0, float}
    Coalesence phase of the binary (in rad).
inclination : {0.0, float}
    Inclination (rad), defined as the angle between the total angular momentum J and the line-of-sight.
delta_f : {None, float}
    The frequency step used to generate the waveform (in Hz).
f_lower : {None, float}
    The starting frequency of the waveform (in Hz).
approximant : {None, str}
    A string that indicates the chosen approximant.
f_ref : {0, float}
    The reference frequency.
phase_order : {-1, int}
    The pN order of the orbital phase. The default of -1 indicates that all implemented orders are used.
spin_order : {-1, int}
    The pN order of the spin corrections. The default of -1 indicates that all implemented orders are used.
tidal_order : {-1, int}
    The pN order of the tidal corrections. The default of -1 indicates that all implemented orders are used.
amplitude_order : {-1, int}
    The pN order of the amplitude. The default of -1 indicates that all implemented orders are used.
f_final : {0, float}
    The ending frequency of the waveform. The default (0) indicates that the choice is made by the respective approximant.

Returns
-------
hplustilde: FrequencySeries
    The plus phase of the waveform in frequency domain.
hcrosstilde: FrequencySeries
    The cross phase of the waveform in frequency domain.
File:      ~/.virtualenvs/pycbc-kombine/lib/python2.7/site-packages/PyCBC-fdc8ff-py2.7.egg/pycbc/waveform/waveform.py
Type:      function
```

**Comment by cdcapano** [comment]
@ahnitz Any word on this? I have some changes to the inference codes that depends on these patches.

**Pull request merged**